packages/plans-grid-next/src/components/comparison-grid/hooks/useCompactViewOrder.ts

```diff
@@ -0,0 +1,87 @@
+import {
+	PlanSlug,
+	URL_FRIENDLY_TERMS_MAPPING,
+	UrlFriendlyTermType,
+	findFirstSimilarPlanKey,
+	isFreePlan,
+} from '@automattic/calypso-products';
+import { WpcomPlansUI } from '@automattic/data-stores';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useEffect } from 'react';
+import { GridPlan } from '../../../types';
+
+type Props = {
+	/**
+	 * Selected plan interval
+	 */
+	intervalType: UrlFriendlyTermType;
+	/**
+	 * The map of displayed plans for the current interval
+	 */
+	gridPlansIndex: { [ key: string ]: GridPlan };
+};
+
+/**
+ *
+ * When intervalType change trickles down to the ComparisonGrid as a prop (intervalType),
+ * we must detect this prop change and transform the current plan order into a set of plans
+ * that align with the currently selected plan interval without causing glitches or flashes.
+ *
+ * This hook serves as the solution. The hook accomplishes two main tasks:
+ * - Resolves the correct compact view plan order, providing it immediately in the current render cycle to prevent glitches.
+ * - Updates the WpcomPlansUI store with the most up-to-date plan order as required.
+ *
+ * ******
+ * @typedef {Object} Param1
+ * @property {UrlFriendlyTermType} intervalType - Selected plan interval
+ * @property {Object.<string, GridPlan>} gridPlansIndex - The map of displayed plans for the current interval
+ * @returns The list of gridPlans sorted by the accurate compact view order.
+ */
+
+function useCompactViewOrder( { intervalType, gridPlansIndex }: Props ): GridPlan[] {
+	const { setComparisonGridCompactViewPlanDisplayOrder } = useDispatch( WpcomPlansUI.store );
+	const compactViewPlanOrder = useSelect(
+		( select ) => select( WpcomPlansUI.store ).getComparisonGridCompactViewDisplayOrder(),
+		[]
+	);
+	const displayedPlanSlugs = Object.keys( gridPlansIndex ) as PlanSlug[];
```

Comment:
There's a bit of indirection throughout the logic that feels unnecessary. Why do you have to work with `plansSlug`s and not with `GridPlan`s directly? All of these conversions and new names aren't necessary.